### ISSUE TITLE
[Bug] Class redefinitions break the Bear :sob: 

### ISSUE THREAD

**Issue by awf** [beartype/beartype#288]
The following code, simulating a notebook scenario where one might redefine some classes by re-running the same cell, roars when it should not (see "Roars" below).
```
from dataclasses import dataclass
from beartype import beartype
from beartype.typing import List, Set, Any, Tuple, Type, List

@beartype
def setminus(s: set, v: Any):
    """
    return s \ {v}
    """
    return set.difference(s, set((v,)))

assert setminus(set((1,3,6)), 3) == set((1, 6))

#### repeat

@beartype
class Expr:
    @property
    def ty(self):
        for ty in (Const, Var, Let, Lambda, Call):
          if isinstance(self, ty): 
              return ty
        assert f"Bad type {self}"

def exprclass(klass, **kwargs):
    return beartype(dataclass(klass, frozen=True,  **kwargs))

@exprclass
class Const(Expr):
    val: Any

@exprclass
class Var(Expr):
    name: str

@exprclass
class Let(Expr):
    var: Var
    val: Expr
    body: Expr

@exprclass
class Lambda(Expr):
    args: List[Var]
    body: Expr

@exprclass
class Call(Expr):
    f: Expr
    args: List[Expr]


def mkvars(s:str) -> Tuple[Var]:
  s = s.replace(' ','')
  return [Var(s) for s in s.split(',')]

foo, x, y, z = mkvars("foo, x, y, z")

a_lam= Lambda([z], Call(Const("boo"), [z]))
call_lam = Call(a_lam, [x])
foo_lam = Lambda([x, y], 
                 Let(z, Call(Const("add"), [x, Const(3.3)]),
                     Call(Const("call_lam_y"), [call_lam, y])))
e = Let(foo, foo_lam, 
        Call(foo, [Const(1.1), Const(2.2)]))

#### repeat

#### repeat

@beartype
class Expr:
    @property
    def ty(self):
        for ty in (Const, Var, Let, Lambda, Call):
          if isinstance(self, ty): 
              return ty
        assert f"Bad type {self}"

def exprclass(klass, **kwargs):
    return beartype(dataclass(klass, frozen=True,  **kwargs))

@exprclass
class Const(Expr):
    val: Any

@exprclass
class Var(Expr):
    name: str

@exprclass
class Let(Expr):
    var: Var
    val: Expr
    body: Expr

@exprclass
class Lambda(Expr):
    args: List[Var]
    body: Expr

@exprclass
class Call(Expr):
    f: Expr
    args: List[Expr]


def mkvars(s:str) -> Tuple[Var]:
  s = s.replace(' ','')
  return [Var(s) for s in s.split(',')]

foo, x, y, z = mkvars("foo, x, y, z")

assert isinstance(z, Expr)  # Passes

a_lam= Lambda([z], Call(Const("boo"), [z])) # Roars
# beartype.roar.BeartypeCallHintParamViolation: Method __main__.Call.__init__() parameter args=[Var(name='z')] violates type hint list[__main__.Expr], as list index 0 item <class "__main__.Var"> "Var(name='z')" not instance of <class "__main__.Expr">.

call_lam = Call(a_lam, [x])
foo_lam = Lambda([x, y], 
                 Let(z, Call(Const("add"), [x, Const(3.3)]),
                     Call(Const("call_lam_y"), [call_lam, y])))
e = Let(foo, foo_lam, 
        Call(foo, [Const(1.1), Const(2.2)]))

#### repeat
```

**Comment by leycec** [comment]
**Gah!** I can sadly confirm this issue on this otherwise wonderfully warm Canadian Sunday. My Sunday is now ruined, I feel bad, and my heart is palpating. This is the high cost of open-source.

While *awesome*, your example is pretty plus-sized. I've reduced that to this minimal reproducible example (MRE) exhibiting the same issue – which I can now wrap my head around. Incidentally, isn't that a *really* weird idiom? The real-world implications of physically wrapping ours heads (which are *not* very wrappable in their natural states) around various household items are concerning. Heads just aren't supposed to do that.

MRE *or* bust:

```python
from beartype import beartype
from beartype.typing import List

class HearMyPlea(object): pass

@beartype
class OdinSaveUs(object):
    def __init__(self, yem: List[HearMyPlea]) -> None:
        pass

class HearMyPlea(object): pass

@beartype
class OdinSaveUs(object):
    def __init__(self, yem: List[HearMyPlea]) -> None:
        pass

OdinSaveUs([HearMyPlea(),])
```

...which raises the similar exception traceback:

```python
Traceback (most recent call last):
  File "/home/leycec/tmp/oapy.py", line 21, in <module>
    OdinSaveUs([HearMyPlea(),])
  File "<@beartype(__main__.OdinSaveUs.__init__) at 0x7fd531462520>", line 33, in __init__
beartype.roar.BeartypeCallHintParamViolation: Method __main__.OdinSaveUs.__init__() parameter
yem=[<__main__.HearMyPlea object at 0x7fd531484350>] violates type hint list[__main__.HearMyPlea], as
list index 0 item <class "__main__.HearMyPlea"> <__main__.HearMyPlea object at 0x7fd531484350> not
instance of <class "__main__.HearMyPlea">.
```

This is *almost* certainly a memoization (i.e., caching) issue. Curiously, this only occurs with nested type hints like `List[HearMyPlea]`; non-nested type hints like `HearMyPlea` are fine, inexplicably. This is why I tug my gray beard in frustration.

Thanks so much for this awesome issue, @awf. I'll get this straightened out straightaway for Cambridge in our upcoming @beartype 0.16.3 patch release. Go, academia! Go! :man_teacher:

**Comment by leycec** [comment]
**Resolved by 0ac4c5c.** If tests pass, <sup>*...come on! my forehead is sweating*</sup> we'll push this out as a new @beartype 0.16.3 release. This release is wonderful. Wonderful things this release does include:

* Optimizing the `@beartype` decorator to protect against redecoration. Previously, `@beartype` inefficiently redecorated user-defined classes that have already been decorated by `@beartype`. Now, `@beartype` efficiently avoids redecorating user-defined classes that have already been decorated by `@beartype`: e.g.,

  ```python
  @beartype  # <-- this now reduces to a noop
  @beartype  # <-- this still does nice stuff
  class MuhClass(...): ...
  ```
* Safeguarding the `@beartype` decorator against **class redefinition** (i.e., redefinition of the same @beartype-decorated class, typically due to hot reloading). Notably, `@beartype` now (in order):

  1. Detects class redefinition.
  1. When detected, clears all **problematic caches** (i.e., @beartype-specific internal caches that could possibly contain the prior definition of the current redefined class).

## tl;dr

It was caching. :sweat_smile:
